orchestrator/src/worker/event_handler/triggers/snos.rs

```diff
@@ -1,176 +1,497 @@
-            .collect();
+    /// Schedules initial jobs when no SNOS jobs have completed yet.
+    ///
+    /// This handles the bootstrap scenario where:
+    /// - No SNOS jobs have completed successfully (middle_limit is None)
+    /// - All blocks in the valid range need to be considered for processing
+    /// - Missing blocks are identified and scheduled within slot limits
+    ///
+    /// # Use Cases
+    /// - Fresh deployment with no processing history
+    /// - Recovery from complete job failure scenarios
+    /// - Initial processing of historical blocks
+    ///
+    /// # Arguments
+    /// * `config` - Application configuration
+    /// * `context` - Mutable scheduling context to update with selected blocks
+    ///
+    /// # Returns
+    /// * `Result<()>` - Success or error from block selection
+    async fn schedule_initial_jobs(&self, config: &Arc<Config>, context: &mut JobSchedulingContext) -> Result<()> {
+        let candidate_blocks = self
+            .get_missing_blocks_in_range(
+                config,
+                context.bounds.lower_limit,
+                context.bounds.upper_limit,
+                context.available_slots,
+            )
+            .await?;
+
+        context.blocks_to_process.extend(candidate_blocks);
+        Ok(())
+    }
 
-        // First add missing blocks (they take priority)
-        let missing_blocks_to_add = missing_blocks.into_iter().take(available_job_slots as usize).collect::<Vec<_>>();
+    /// Schedules jobs for a specific block range if processing conditions are met.
+    ///
+    /// This method handles range-specific job scheduling with several safety checks:
+    ///
+    /// # Range Validation Cases
+    /// - **Invalid Range** (start > end): Skip processing, log debug message
+    /// - **Empty Range** (start == end == 0): Skip processing (nothing to do)
+    /// - **Valid Range**: Proceed with job scheduling
+    ///
+    /// # Processing Logic
+    /// 1. Validate range boundaries and skip if invalid
+    /// 2. Query database for missing blocks in the range
+    /// 3. Respect slot limits when selecting blocks
+    /// 4. Update context with selected blocks and remaining slots
+    ///
+    /// # Arguments
+    /// * `config` - Application configuration
+    /// * `context` - Mutable scheduling context (updated with blocks and slots)
+    /// * `start` - Starting block number (inclusive)
+    /// * `end` - Ending block number (inclusive)
+    /// * `range_name` - Human-readable name for logging ("first half", "second half")
+    ///
+    /// # Returns
+    /// * `Result<()>` - Success or error from block selection
+    async fn schedule_jobs_for_range(
+        &self,
+        config: &Arc<Config>,
+        context: &mut JobSchedulingContext,
+        start: u64,
+        end: u64,
+        range_name: &str,
+    ) -> Result<()> {
+        // Skip if range is invalid or empty
+        if start > end || (start == end && end == 0) {
```

Comment:
> [!WARNING]
> _**Invalid condition**_

Wouldn't `start == end` also qualify as an empty range? Right now this can be simplified to `start > end || end == 0`, since if `start != end` and `end == 0` then `start > end` since `start >= 0`.

```suggestion
        if start > end || start == end || end == 0 {
```

Reply:
Great catch !
Added.

Reply:
_Actually this can be simplified down to `start > end || start == end` but at this point I'm nitpicking :sweat_smile:_